Add update service wizard, support readonly catalog parameters.

jeff-phillips-18 opened the pull request:
Adding an update service wizard as well as updating the catalog-parameters component to provide for read-only form and horizontal layout and the ability to hide the current values.

TODO:
- [x]  Get update service instance to work correctly, currently setting the spec.updateRequests value is not working as expected
- [x]  Test with service classes with multiple plans 
- [x]  Test with instanceUpdateParameterSchema set.

rhamilto commented on dist/less/catalogParameters.less:
```diff
@@ -0,0 +1,9 @@
+  input.form-control[disabled] {
+    color: @text-color;
+    cursor: default;
+    background-color: transparent;
```
Nit: alpha

rhamilto commented on dist/less/catalogParameters.less:
```diff
@@ -0,0 +1,9 @@
+.catalog-parameters {
+  input.form-control[disabled] {
```
Nit:  nesting is unnecessary since there is only one rule

rhamilto commented on src/components/update-service/update-service-review.html:
```diff
@@ -0,0 +1,46 @@
+      <span class="sr-only">Error</span>
+      <div class="results-message">
+        <h3>
+          Failed to updaate <strong>{{$ctrl.serviceName}}</strong> failed to provision in <strong>{{$ctrl.project | displayName}}</strong>.
```
Typo: "updaate"

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,338 @@
+    var serviceInstancesVersion: any = this.APIService.getPreferredVersion('serviceinstances');
+    // TODO: This is not working
+    updateInstance.spec.updateRequests = updateInstance.spec.updateRequests ? updateInstance.spec.updateRequests + 1 : 1;
+    this.DataService.update(serviceInstancesVersion, _.get(this.ctrl.serviceInstance, 'metadata.name'), updateInstance, this.context).then((updated: any) => {
```
We shouldn't continue updating the instance if creating the secret failed.

jeff-phillips-18 commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,338 @@
+    var promises: any = [];
+
+    _.each(_.get(this.ctrl.serviceInstance, 'spec.parametersFrom'), (parametersSource) => {
+      promises.push(this.DataService.get("secrets", _.get(parametersSource, 'secretKeyRef.name'), this.context).then((secret: any) => {
```
Would there ever be? While it would do it twice, it would only override its own values.

spadgett replied:
I think there can be more than on as long as the keys of the JSON objects are unique in each.

spadgett replied:
Could they reference two different keys in the same secret, though?

jeff-phillips-18 commented:
I believe I have addressed @rhamilto and @spadgett comments.

This still has not been fully tested with multiple plans and/or editable plans.

jeff-phillips-18 commented:
Tested updating the service plan. 
Fixed issue where plans were not showing as radios. 
Fixed watching for the update success/fail condition.
Fixed title for the wizard.

jeff-phillips-18 commented:
All Todo's completed. Ready for final reviews.

spadgett commented on src/components/select-plan/selectPlan.html:
```diff
@@ -0,0 +1,25 @@
+      No Plans Available
+    </h1>
+    <p>
+      There are no plans currently available for this service class.
```
nit: "service class" is the k8s name, but we don't use it in the UI. So maybe just, "There are no plans currently available for this service."

spadgett commented on src/components/update-service/update-service-results.html:
```diff
@@ -0,0 +1,45 @@
+<div class="order-service-config">
+  <div ng-if="!$ctrl.error">
+    <div ng-if="!$ctrl.orderComplete">
```
One div?

```html
<div ng-if="!$ctrl.error && !ctrl.orderComplete">
```

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,355 @@
+    this.ctrl.configPageShown = false;
+
+    var updateProperties: any = _.get(this.ctrl.parameterSchema, 'properties');
+    this.configStep.hidden = _.size(updateProperties) < 1;
```
Just confirming it's not possible to show the dialog without plans or parameter schema. Are we checking for schema `properties` in origin-web-console?

jeff-phillips-18 replied:
We are. Should I add a safe guard here? What should we do? Close the panel? Display a message?

spadgett replied:
I think it's OK, just wanted to confirm.

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,355 @@
+
+    this.validityWatcher = this.$scope.$watch("$ctrl.parameterSchema", () => {
+      var updateProperties: any = _.get(this.ctrl.parameterSchema, 'properties');
+      this.configStep.hidden = _.size(updateProperties) < 1;
```
Might pull these two lines into a helper since it's repeated from above.

jeff-phillips-18 replied:
Actually all of this can just be done whenever we update the parameterSchema at that time (no need to watch). Thanks.

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,355 @@
+      }
+    });
+
+    // Add a new secret if any secret parameters were updated
```
What happens if a new parameter is added wasn't already set in a secret or in the instance?

spadgett replied:
Also are we handling the case where we're clearing an existing value properly?

jeff-phillips-18 replied:
Clearing works properly, if it was in the parameters it will be removed. if it was in a secret, the secret gets removed from its list and a new secret is added.

Still working on checking for new parameters

jeff-phillips-18 replied:
Added code to add any new parameters to a new secret.

jeff-phillips-18 commented on src/components/catalog-parameters/catalog-parameters.controller.ts:
```diff
@@ -35,22 +63,72 @@ export class CatalogParametersController implements angular.IController {
       }
     };
+    if (this.ctrl.isHorizontal) {
+      angular.extend(this.ctrl.parameterFormDefaults.formDefaults, {
```
If isHorizontal changes `setupFormDefaults` is re-called and it would go back to stacked.

spadgett replied:
OK yeah I see now. Thanks

jeff-phillips-18 commented on src/components/catalog-parameters/catalog-parameters.controller.ts:
```diff
@@ -35,22 +63,72 @@ export class CatalogParametersController implements angular.IController {
-      this.ctrl.parameterForm = this.cloneParameterForm(this.ctrl.parameterFormDefinition) || ['*'];
+  private setupReadonlySchema() {
+    if (!this.ctrl.parameterSchema || !this.ctrl.readOnly) {
+      return;
```
When readonly it uses the readonlyParameterSchema. so this is covered. However, we do need to update the parameter form. I will add the readonly changed check for that.

spadgett commented:
Were you able to test other read-only input types like checkboxes?

jeff-phillips-18 commented:
Everything but array types, thanks for the reminder. I'll fake it and test it.

jeff-phillips-18 commented:
Addressed all comments thus far, still testing different types in readonly forms

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,348 @@
+    _.each(this.secrets, (secret: any) => {
+      var originalSecretParameters: any = JSON.parse(this.SecretsService.decodeSecretData(secret.data).parameters);
+      var updateSecretParameters: any = _.pick(parameterData, _.keys(originalSecretParameters));
+      newParameters = _.omit(newParameters, _.keys(originalSecretParameters));
```
I don't think this handles property names with a `.` in them correctly. It will treat them like JSON object paths.

Maybe (untested)

```js
var keys = _.map(originalSecretParameters, (value: any, key: string) => {
  // Make sure properties with a `.` in the name aren't treated as paths.
  return [ key ];
};
var updateSecretParameters: any = _.pick(parameterData, keys);
newParameters = _.omit(newParameters, keys);
```

jeff-phillips-18 replied:
Tested and this is working as is when there is a . in the key. I can update if you aren't comfortable though.

jeff-phillips-18 replied:
I went with your suggestion, cleaner to get the keys just once anyhow.

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,348 @@
+        // Remove the reference to this secret
+        updateInstance.spec.parametersFrom = _.reject(updateInstance.spec.parametersFrom, function (parameters: any) {
+          return _.get(parameters, 'secretKeyRef.name') === secret.metadata.name;
+        });
```
nit: could use shorthand here:

```js
_.reject(updateInstance.spec.parametersFrom, { secretKeyRef: { name: secret.metadata.name } });
```

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,348 @@
+    // Update the instance
+    var serviceInstancesVersion: any = this.APIService.getPreferredVersion('serviceinstances');
+    var isServiceInstanceReady = this.$filter('isServiceInstanceReady');
+    var isBindingFailed = this.$filter('isBindingFailed');
```
Should use `isServiceInstanceFailed`

spadgett commented on src/components/update-service/update-service.controller.ts:
```diff
@@ -0,0 +1,348 @@
+    var serviceInstancesVersion: any = this.APIService.getPreferredVersion('serviceinstances');
+    var isServiceInstanceReady = this.$filter('isServiceInstanceReady');
+    var isBindingFailed = this.$filter('isBindingFailed');
+    var bindingFailedMessage = this.$filter('bindingFailedMessage');
```
serviceInstanceFailedMessage

jeff-phillips-18 commented:
Tested with different types (including array and object) found some necessary modifications.

jeff-phillips-18 commented:
Rebased. @rhamilto Any further comments?

jeff-phillips-18 commented:
Updated the update service wizard to use the new wizard-title parameter.
